### DISCUSSION TITLE
How to manage roles of users for a model

### DISCUSSION THREAD

**Discussion by nilskretschmer** [filamentphp/filament#12606] [Help]
### Package

Panel builder

### Package Version

v3.2.71

### How can we help you?

Hi,

I am trying to set up Filament to manage the roles of users in my application. Roles can also be attached to some models (in this case a `Project`-Model). The relationships are set up like this:

- Project
```
    public function roles(): MorphMany
    {
        return $this->MorphMany(Role::class, 'model');
    }
```
- Roles
```
    public function model(): MorphTo
    {
        return $this->morphTo();
    }

    public function users(): BelongsToMany
    {
        return $this->belongsToMany(User::class);
    }
```
- Users
```
    public function roles(): BelongsToMany
    {
        return $this->belongsToMany(Role::class);
    }
    public function projects(): HasManyDeep
    {
        return $this->hasManyDeepFromReverse((new Project)->users());
    }
```

As you can see I also use Staudenmeir's `HasManyDeep`-Package.

With this setup I can attach a User to a Role that is also attached to a Project.

Now I am trying to attach/detach users to a project through their roles. A user can either be an `admin` or an `user` of a project (through the role).

For this I created a `UsersRelationManager`:

```
<?php

namespace App\Filament\Resources\ProjectResource\RelationManagers;

use Filament\Forms;
use Filament\Forms\Form;
use Filament\Resources\RelationManagers\RelationManager;
use Filament\Tables;
use Filament\Tables\Table;
use Illuminate\Database\Eloquent\Builder;

class UsersRelationManager extends RelationManager
{
    protected static string $relationship = 'users';

    public function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\Select::make('roles')                    
                    ->relationship(titleAttribute: 'name', modifyQueryUsing: function (Builder $query, RelationManager $livewire) {
                        $query->whereMorphedTo('model', $livewire->getOwnerRecord());
                    })
            ]);
    }

    public function table(Table $table): Table
    {
        return $table
            ->recordTitleAttribute('username')
            ->columns([
                Tables\Columns\TextColumn::make('username'),
                Tables\Columns\TextColumn::make('project role')->getStateUsing(function ($record, RelationManager $livewire) {
                    return $this->userRole($record, $livewire->ownerRecord);
                }),
            ])
            ->filters([
                //
            ])
            ->headerActions([
/*                 Tables\Actions\AttachAction::make(), */
            ])
            ->actions([
/*                 Tables\Actions\DetachAction::make(), */
                Tables\Actions\EditAction::make()->before(function ($record, RelationManager $livewire) {
                    unset($record->roles);
                }),
            ])
            ->bulkActions([
                Tables\Actions\BulkActionGroup::make([
/*                     Tables\Actions\DetachBulkAction::make(), */
                ]),
            ]);
    }

    public function userRole($user, $project)
    {
        if ($user->hasRole('admins', $project)) {
            return 'admin';
        } else {
            return 'user';
        }
    }
}

```

Which looks like this in the frontend:

![grafik](https://github.com/filamentphp/filament/assets/24235417/00ad5502-50e7-4372-b4ad-307f45a5f0a3)

If I now edit a user I get a nice form modal with the current role in the select box:
![grafik](https://github.com/filamentphp/filament/assets/24235417/42450296-f1fa-43bc-ac52-1639c1d86ded)

The problem now:
If I use the `Select`-Field in the form without `multiple()` I get this exception:
`Add fillable property [roles] to allow mass assignment on [App\Models\User].`

I think this is because Filament determines the type of relationship based on the `multiple`-option and so assumes a BelongsTo and not a BelongsToMany-Relation. If I add `multiple` I have a select field for multiple roles - but I do not want to achieve that. In this case the UI would be misleading. I only want a user to have one of the roles in the project (but he can have different roles on other projects, thus making it many-to-many).

But the even bigger problem is that the user is only attached to that specific role that was selected and DETACHED FROM ALL OTHER ROLES. I went through Filament's documentation but I did not find any way to handle the attaching/detaching. It seems like Filament is calling Laravel's `sync`-method instead of attaching/detaching the user from roles?

**Comment by makhsam** [comment]
Adding `->multiple()` to the select field, solved the issue for me. Thanks!
